File: ros-beginners/rospy_simple_service_client/ros/script/simple_server.py
#!/usr/bin/env python

"""
This script is a simple python node which advertizes ros service
server rospy_word_count
"""
#-*- encoding: utf-8 -*-
__author__ = 'mohammad wasil'

import rospy
from ros_custom_msgs.srv import CountWords, CountWordsResponse

def server_callback(req):
  rospy.loginfo("Received request: %s", req.words.data)
  
  words = req.words.data.split(" ")
  resp = CountWordsResponse()

  if req.count_articles.data:
    english_articles = ['a', 'an', 'the'] 
    resp.num_articles.data = len([word for word in words if word in english_articles])

  resp.num_words.data = len(words) - resp.num_articles.data

  return resp

if __name__ == '__main__':
  rospy.init_node("rospy_simple_server_node", anonymous=False)
  rospy.Service("rospy_word_count", CountWords, server_callback)  
  rospy.spin()

File: ros-beginners/roscpp_simple_service_client/ros/src/simple_server.cpp
#include <ros/ros.h>
#include <ros_custom_msgs/CountWords.h>
#include <boost/algorithm/string.hpp>

bool serverCallback(ros_custom_msgs::CountWordsRequest &req,
                    ros_custom_msgs::CountWordsResponse &resp)
{
  ROS_INFO_STREAM("Request received: "<<req.words.data);
  
  std::vector<std::string> words;
  boost::split(words, req.words.data, boost::is_any_of("\t "));
  
  std::vector<std::string> english_articles = {"a", "an", "the"};
  int num_of_articles = 0;
  for (auto const& word: words)
  {
    std::string lower_word = boost::algorithm::to_lower_copy(word);
    if (std::find(english_articles.begin(), english_articles.end(), 
        lower_word ) != english_articles.end())
    {
      num_of_articles ++;
    }
  }
  
  resp.num_articles.data = num_of_articles;
  resp.num_words.data = words.size() - num_of_articles;
  
  return true;
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "roscpp_simple_server_node");
  ros::NodeHandle nh;
  ros::ServiceServer service = nh.advertiseService("roscpp_word_count", serverCallback);
  ros::spin();
  
  return 0;
}
